### PR TITLE
[EraVM][MC] Adjust relocation type when the symbol points to an instructions section.

### PR DESCRIPTION
**Pull request by PavelKopyl**
# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: lld/ELF/Arch/EraVM.cpp [hunk 1]
```diff
@@ -52,17 +52,37 @@ void EraVM::relocate(uint8_t *loc, const Relocation &rel, uint64_t val) const {
...
+  //     add   @.OUTLINED_FUNCTION_RET0[0], r0, stack-[1]
+  //     jump  @OUTLINED_FUNCTION_0
+  //   .OUTLINED_FUNCTION_RET0:
+  //     jump.ge @.BB1_16
+  //     jump  @.BB1_23
+  //
+  // In such cases the actual relocation type should be R_ERAVM_16_SCALE_8.
+  // TODO: check if this can be done properly done on the LLVM MC layer?
+  auto *sec = cast<InputSection>(cast<Defined>(rel.sym)->section);
+  if (sec->name == ".text" && rel.type == R_ERAVM_16_SCALE_32)
```
Comment (vladimirradosavljevic) [inline lld/ELF/Arch/EraVM.cpp]: Instead of checking section name, maybe it is safer to check if section flags have `SHF_EXECINSTR`?
Also, regarding fix in the LLVM MC layer, maybe we can do something similar to [this](https://github.com/matter-labs/era-compiler-llvm/blob/9e9009229ea05aca97df5695933260cb2d78d010/llvm/lib/Target/BPF/MCTargetDesc/BPFELFObjectWriter.cpp#L54)?

Comment (PavelKopyl) [reply lld/ELF/Arch/EraVM.cpp]: Hey @vladimirradosavljevic, thank you for the link! Will check it.

Comment (PavelKopyl) [reply lld/ELF/Arch/EraVM.cpp]: Implemented the relocation type adjustment in the MC layer.

**Review by vladimirradosavljevic — APPROVED** [review]
LGTM with title changed.
Please wait with merging for 1 day, as others may want to look into this.

File: llvm/lib/Target/EraVM/MCTargetDesc/EraVMELFObjectWriter.cpp [hunk 2]
```diff
@@ -39,8 +39,32 @@ class EraVMELFObjectWriter : public MCELFObjectTargetWriter {
...
+      //   .OUTLINED_FUNCTION_RET0:
+      //     jump.ge @.BB1_16
+      //     jump  @.BB1_23
+      //
+      // Here the .OUTLINED_FUNCTION_RET0[0] represents the code offset,
+      // measured in 8-byte units.
+      // In such cases the actual relocation type should be R_ERAVM_16_SCALE_8.
+      if (const MCSymbolRefExpr *A = Target.getSymA()) {
+        const MCSymbol &Sym = A->getSymbol();
+        assert(Sym.isDefined());
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EraVM/MCTargetDesc/EraVMELFObjectWriter.cpp:57]: Is there any case where we would create relocation here for undefined symbol?

Comment (PavelKopyl) [reply llvm/lib/Target/EraVM/MCTargetDesc/EraVMELFObjectWriter.cpp:57]: Such cases shouldn't exist (to the beast of my knowledge), because in the current design we have only one compilation unit and all the symbols must be defined. That's why I put an `assert` here instead of `if()`.
I guess we will get undefined symbols once we add libraries support, but they will have another fix-up/relocation type.

Comment (vladimirradosavljevic) [reply llvm/lib/Target/EraVM/MCTargetDesc/EraVMELFObjectWriter.cpp:57]: I had the same thinking, just wanted to confirm. Thanks!

**Review by akiramenai — APPROVED** [review]

**Comment by PavelKopyl** [comment]
@akiramenai , @vladimirradosavljevic , please don't merge it because the PR is made agains `kpv-eravm-init-code` branch. I can make it against `main`, but this will slightly change the test.

**Comment by PavelKopyl** [comment]
I updated the test, by removing `r0` operand of ` ret.panic.to_label` instruction and made the PR against the `main` branch.

**Comment by akiramenai** [comment]
So now it can be merged. Am I right?

**Comment by akiramenai** [comment]
CI run: https://github.com/matter-labs/era-compiler-llvm/actions/runs/10043224307

**Comment by PavelKopyl** [comment]
> CI run: https://github.com/matter-labs/era-compiler-llvm/actions/runs/10043224307

Thank you for running the CI, I also wanted to run it again.  But it looks like there is something wrong with it now.

**Comment by PavelKopyl** [comment]
> So now it can be merged. Am I right?

Yep, once the CI is passed.

**Comment by akiramenai** [comment]
Rebased to check that benchmark are not affected.
